Reorganize opaque ports deep tests

krzysztofdrys opened the pull request:
There are three major changes in this PR.

1. The are two new scenarios introduced. Previously there was only one
   scenario: Slow Cooker "calling" service by it's name. Two new
   scenarios are SC using service IP or pod IP as a target.
2. Each scenario is now structured as table test.
3. For each scenario there is now assertion on the outbound TCP traffic.

Signed-off-by: Krzysztof Dryś <krzysztofdrys@gmail.com>

<!--  Thanks for sending a pull request!

If you already have a well-structured git commit message, chances are GitHub
set the title and description of this PR to the git commit message subject and
body, respectively. If so, you may delete these instructions and submit your PR.

If this is your first time, please read our contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md

The title and description of your Pull Request should match the git commit
subject and body, respectively. Git commit messages are structured as follows:

```
Subject

Problem

Solution

Validation

Fixes #[GitHub issue ID]

DCO Sign off
```

Example git commit message:

```
Introduce Pull Request Template

GitHub's community guidelines recommend a pull request template, the repo was
lacking one.

Introduce a `PULL_REQUEST_TEMPLATE.md` file.

Once merged, the
[Community profile checklist](https://github.com/linkerd/linkerd2/community)
should indicate the repo now provides a pull request template.

Fixes #3321

Signed-off-by: Jane Smith <jane.smith@example.com>
```

Note the git commit message subject becomes the pull request title.

For more details around git commits, see the section on Committing in our
contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md#committing
-->

krzysztofdrys commented:
This PR ended up much bigger than I initially intended. Sorry about that!

mateiidavid commented on test/integration/deep/opaqueports/assertions_test.go:
```diff
@@ -0,0 +1,75 @@
+)
+
+var (
+	tcpMetricRE = regexp.MustCompile(
```
A parsing library for those metrics would be cool. It'd be in favour of having an util in TestHelper that can help us make sense of the proxy metrics; definitely shouldn't be scoped to this PR though.

I'd be in favour of rolling those regex assertions into one of the test files. Very subjective, but think at the very least we should remove `_test` from its name so the test runner doesn't try to compile/run it, or whatever it is that Go's runner does in those situations.

krzysztofdrys replied:
I submitted this issue: https://github.com/linkerd/linkerd2/issues/8250

krzysztofdrys replied:
As for:
> I'd be in favour of rolling those regex assertions into one of the test files. Very subjective, but think at the very least we should remove _test from its name so the test runner doesn't try to compile/run it, or whatever it is that Go's runner does in those situations.

I think Go compiler will always compile it when running tests, regardless of `_test` suffix. What is important is that these regex are required for running tests.

kleimkuhler replied:
There are no tests in here, so I think it makes more to call this something like `util.go` for now since it will already be in the `opaqueports` integration test package. Even if just for the developer, it's easier to realize there are no actual tests in here; I would expect a `..._test.go` file to have tests in it.

That being said, it's only used by `opaque_ports_test.go` so I'm also fine just moving things back to a single file. I don't think introducing a separate utility file that is still only used by one other file to be significantly clearer.

krzysztofdrys replied:
@kleimkuhler How about calling this file `assertions.go`? Or `checks.go`? 

Since this file contains checks/assertions for metrics, I would like to avoid a generic name like "util".

krzysztofdrys replied:
Friendly ping @kleimkuhler

krzysztofdrys replied:
Rename done in https://github.com/linkerd/linkerd2/pull/8225/commits/8d9fc2f0a181826356a144bc48851bcc6d568759

mateiidavid reviewed (commented):
Looking good so far, left a few comments. I think what I'd like to see is whether we can organise these tests in such a way that they can:

* Make use of the shared functions you've introduced;
* Be in different test packages so that they can be run in parallel -- I think it'd help a bit with the runtime if we split the tests between direct (e.g podIP) and logical targets.

krzysztofdrys commented on test/integration/deep/opaqueports/opaque_ports_test.go:
Good catch! Done in https://github.com/linkerd/linkerd2/pull/8225/commits/ee09fd150150d2665b97d0d222b4a4dd25235197

krzysztofdrys commented on test/integration/deep/opaqueports/opaque_ports_test.go:
The difference between these two is that one uses service name and the other uses service IP. So if we have:
```
> kubectl get svc -n linkerd-opaque-ports-called-by-service-name-test
NAME                          TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)                  AGE
svc-opaque-service   ClusterIP   10.43.223.137   <none>        8080/TCP                 11
```
in one test SC will call `svc-opaque-service:8080` and the other will call `10.43.223.137:8080`. 

If you say that `CalledByServiceIP` is redundant, then I am willing to remove it. However, my personal comment is that it felt natural for me to have this test, as another way a pod can call a service. And when you say:

>Whether the target is named or unnamed (e.g IP) shouldn't make a difference 

my first instinct is that this is what tests are for: to make sure we are right about our assumptions.

Having said all that, I don't know that much about linkerd and name resolution. I was a little surprised that DNS is involved in this scenario. 🙃 So if you feel like this test is redundant and this stuff is tested somewhere else, then I am all willing to remove it.

krzysztofdrys replied:
ok, I was thinking about it. What you are saying is that resolution of `svc-opaque-service` to `10.43.223.137` will happen before app will "talk to" linkerd sidecar, right?  If I get this right, I think we should get rid of this test.

mateiidavid replied:
@krzysztofdrys 

> What you are saying is that resolution of svc-opaque-service to 10.43.223.137 will happen before app will "talk to" linkerd sidecar, right?

Yup. This wouldn't make a difference in how opaque ports are handled. We can just use the fqn since it's easier to read in tests (and we don't need any logic to get the service's IP).

krzysztofdrys commented:
@mateiidavid following you suggestion, I have reorganised the tests. I removed tests using service ip and renamed the remaining two. https://github.com/linkerd/linkerd2/pull/8225/commits/dd36197f56e1110169327e20742b930a071eafaa

mateiidavid reviewed (commented):
@krzysztofdrys looks great! Nice refactor, sorry it took me long to reply. The only left thing left to address, from my perspective, is the blocking logic. I'd rather we move to `WaitRollout` to pass in the test helper and have nicer error handling but that's about it :D

krzysztofdrys commented on test/integration/deep/opaqueports/assertions.go:
```diff
@@ -0,0 +1,75 @@
+	)
+)
+
+func hashNoOutbondHTTPRequest(metrics string) error {
```
Fixed in https://github.com/linkerd/linkerd2/pull/8225/commits/704e2189c76cdb5b1c14ff357769aa5982bb2586

krzysztofdrys commented on test/integration/deep/opaqueports/opaque_ports_test.go:
Done in https://github.com/linkerd/linkerd2/commit/e69e1d47ac5d889fa92ddaf66ac1e7949bcbd957

I decided to keep `waitForAppDeploymentReady` because I like that it is short and the name explains what it does.

mateiidavid reviewed (approved):
@krzysztofdrys thanks! 🚢 📦

kleimkuhler commented on test/integration/deep/opaqueports/opaque_ports_test.go:
```suggestion
			testutil.AnnotatedFatal(t, "failed to fetch Pod IPs", err)
```

krzysztofdrys replied:
Fixed in https://github.com/linkerd/linkerd2/pull/8225/commits/a2f7944f42a2151185fe3109ed456f3b05df9449 I went for lower case, because it seems more natural to me in this case.

kleimkuhler reviewed (commented):
Hey @krzysztofdrys sorry about my comments on here and then not much else from me until now. Overall this looks good aside from the one failure message that I commented on.

The one larger thing I'll say is that `TestOpaquePortsCalledByServiceTarget` and `TestOpaquePortsCalledByPodTarget` differ in the setups for `deployClients`, but after that they are identical. They create the same `tests`, and for each `tc` assert the same things.

This makes sense as they only _should_ differ in what the clients actually target. We could probably de-dup some of this by having each of those tests call into a shared helper that is responsible for creating `tests` and running through the assertions for each test.

Does that make sense? I'm also happy to approve this now and do it as a follow-up since it's not a blocker for me—just an opportunity to simplify `opaque_ports_test.go` a little.

krzysztofdrys commented:
@kleimkuhler I added a helper function. It executes the tests. I hope it is better now. 

Tests targeting services and pods have slightly different assertions. The difference is whether we expect authority in the metrics.  So I don't see much room for more deduplication.
